src/lib/FreeC/Pass/EtaConversionPass.hs

```diff
@@ -51,45 +97,142 @@
+
+import           Data.Maybe                     ( fromMaybe
+                                                , fromJust
+                                                )
 
 import           FreeC.Environment
 import           FreeC.Environment.Fresh
+import           FreeC.Environment.Entry
 import           FreeC.IR.SrcSpan
 import           FreeC.IR.Subterm
 import qualified FreeC.IR.Syntax               as IR
 import           FreeC.Monad.Converter
 import           FreeC.Pass
+-- temporary import; TODO: move this pass before the TypeSignaturePass.
+import           FreeC.Pass.TypeSignaturePass   ( splitFuncType )
+
 
 -- | Applies η-conversions to the right-hand sides of all function declarations
 --   in the given module until all function and constructor applications are
 --   fully applied.
 etaConversionPass :: Pass IR.Module
 etaConversionPass ast = do
-  funcDecls' <- mapM etaConvertFuncDecl (IR.modFuncDecls ast)
+  funcDecls' <- etaConvertFuncDecls (IR.modFuncDecls ast) []
   return ast { IR.modFuncDecls = funcDecls' }
 
 -------------------------------------------------------------------------------
 -- Function declarations                                                     --
 -------------------------------------------------------------------------------
 
--- | Applies 'etaConvertExpr' to the right-hand side of the given function
---   declaration.
+-- | Makes sure that all occurring functions are fully applied
+--   by calling 'etaConvertFuncDecl' on each of them.
+-- 
+--   If the type of a function declaration changes due to a top-level 
+--   η-conversion, the procedure is performed recursively on all 
+--   previously converted function declarations. 
+--   This ensures that all functions, including mutually-recursive 
+--   functions, are fully applied correctly. 
+
+--   The function's first argument represents the function 
+--   declarations yet to be converted.
+--   The function's second argument is an accumulator for already 
+--   converted functions that is needed in case they must be 
+--   re-converted recursively. 
+etaConvertFuncDecls :: [IR.FuncDecl] -> [IR.FuncDecl] -> Converter [IR.FuncDecl]
+etaConvertFuncDecls []         newFuncDecls = return newFuncDecls
+etaConvertFuncDecls (fd : fds) newFuncDecls = do
+  newFuncDecl <- etaConvertFuncDecl fd
+  if IR.funcDeclReturnType newFuncDecl /= IR.funcDeclReturnType fd
+    then etaConvertFuncDecls (newFuncDecls ++ (newFuncDecl : fds)) []
+    else etaConvertFuncDecls fds (newFuncDecls ++ [newFuncDecl])
+
+-- | Applies appropriate η-conversions to a function declaration.
+--
+-- Depending on the presence or absence of missing top-level arguments,
+-- the function uses 'etaConvertTopLevel' or 'etaConvertExpr' to ensure 
+-- all functions and constructors on the right-hand side are fully applied. 
+-- The missing top-level arguments are also added to the left-hand
+-- side of the declaration and the function's type and the environment
+-- are updated accordingly. 
```

Comment:
This paragraph should be aligned with the subject line. There is another line in 207 that needs to be aligned.
After aligning the comments, make sure that the 80 character limit is still met. Some lines are still slightly too long right now.

```suggestion
--   Depending on the presence or absence of missing top-level arguments,
--   the function uses 'etaConvertTopLevel' or 'etaConvertExpr' to ensure 
--   all functions and constructors on the right-hand side are fully applied. 
--   The missing top-level arguments are also added to the left-hand
--   side of the declaration and the function's type and the environment
--   are updated accordingly. 
```